File: app/src/main/java/com/huanchengfly/tieba/post/fragments/MessageFragment.java
package com.huanchengfly.tieba.post.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.tabs.TabLayout;
import com.huanchengfly.tieba.api.TiebaApi;
import com.huanchengfly.tieba.api.models.MessageListBean;
import com.huanchengfly.tieba.api.retrofit.exception.TiebaException;
import com.huanchengfly.tieba.post.MainActivity;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.adapters.MessageListAdapter;
import com.huanchengfly.tieba.post.adapters.TabViewPagerAdapter;
import com.huanchengfly.tieba.post.components.MyLinearLayoutManager;
import com.huanchengfly.tieba.post.components.dividers.CommonDivider;
import com.huanchengfly.tieba.post.interfaces.Refreshable;
import com.huanchengfly.tieba.post.utils.DisplayUtil;
import com.huanchengfly.tieba.post.utils.ThemeUtil;
import com.huanchengfly.tieba.post.utils.Util;
import com.huanchengfly.tieba.widgets.theme.TintImageView;

import org.jetbrains.annotations.NotNull;

import butterknife.BindView;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MessageFragment extends BaseFragment implements View.OnClickListener, Refreshable, TabLayout.OnTabSelectedListener {
    public static final int TYPE_REPLY_ME = 0;
    public static final int TYPE_AT_ME = 1;

    public static final String TAG = MessageFragment.class.getSimpleName();

    private static final String PARAM_TYPE = "type";
    @BindView(R.id.fragment_message_tab)
    TabLayout tabLayout;
    private MessageListHelper replyMe;
    private MessageListHelper atMe;
    private int type;

    public MessageFragment() {
    }

    public static MessageFragment newInstance(int type) {
        MessageFragment fragment = new MessageFragment();
        Bundle bundle = new Bundle();
        bundle.putInt(PARAM_TYPE, type);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onFragmentVisibleChange(boolean isVisible) {
        if (isVisible) {
            refreshIfNeed();
        }
    }

    @Override
    public void onFragmentFirstVisible() {
        refreshIfNeed();
    }

    public int getType() {
        return type;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args != null) {
            type = args.getInt(PARAM_TYPE, TYPE_REPLY_ME);
        }
    }

    @Override
    int getLayoutId() {
        return R.layout.fragment_message;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ViewPager viewPager = view.findViewById(R.id.fragment_message_vp);
        TintImageView searchBtn = view.findViewById(R.id.search_btn);
        if (getAttachContext() instanceof MainActivity) {
            searchBtn.setVisibility(View.VISIBLE);
        } else {
            searchBtn.setVisibility(View.INVISIBLE);
        }
        searchBtn.setOnClickListener(this);
        TabViewPagerAdapter viewPagerAdapter = new TabViewPagerAdapter();
        replyMe = new MessageListHelper(getAttachContext(), TYPE_REPLY_ME);
        atMe = new MessageListHelper(getAttachContext(), TYPE_AT_ME);
        viewPagerAdapter.addView(replyMe.getContentView(), getAttachContext().getString(R.string.title_reply_me));
        viewPagerAdapter.addView(atMe.getContentView(), getAttachContext().getString(R.string.title_at_me));
        viewPager.setAdapter(viewPagerAdapter);
        tabLayout.setupWithViewPager(viewPager);
        tabLayout.addOnTabSelectedListener(this);
        viewPager.setCurrentItem(type, false);
    }

    @Override
    public boolean hasOwnAppbar() {
        return true;
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.search_btn) {
            if (getAttachContext() instanceof MainActivity) {
                ((MainActivity) getAttachContext()).openSearch();
            }
        }
    }

    @Override
    public void onRefresh() {
        switch (tabLayout.getSelectedTabPosition()) {
            case 0:
                if (isFragmentVisible()) {
                    replyMe.refresh();
                } else {
                    replyMe.dataBean = null;
                }
                break;
            case 1:
                if (isFragmentVisible()) {
                    atMe.refresh();
                } else {
                    atMe.dataBean = null;
                }
                break;
        }
    }

    private void refreshIfNeed() {
        switch (tabLayout.getSelectedTabPosition()) {
            case 0:
                if (replyMe.needLoad()) {
                    replyMe.refresh();
                }
                break;
            case 1:
                if (atMe.needLoad()) {
                    atMe.refresh();
                }
                break;
        }
    }

    @Override
    public void onTabSelected(TabLayout.Tab tab) {
        refreshIfNeed();
    }

    @Override
    public void onTabUnselected(TabLayout.Tab tab) {
    }

    @Override
    public void onTabReselected(TabLayout.Tab tab) {
        refreshIfNeed();
    }

    class MessageListHelper {
        private View contentView;
        private SwipeRefreshLayout swipeRefreshLayout;
        private RecyclerView recyclerView;
        private MessageListAdapter adapter;
        private int type;
        private int page;
        private MessageListBean dataBean;

        MessageListHelper(Context context, int type) {
            this.type = type;
            if (this.type != TYPE_REPLY_ME && this.type != TYPE_AT_ME) {
                throw new IllegalArgumentException("参数不正确");
            }
            contentView = Util.inflate(context, R.layout.fragment_message_list);
            if (contentView == null) {
                throw new NullPointerException("引入的布局为空");
            }
            recyclerView = contentView.findViewById(R.id.fragment_message_recycler_view);
            swipeRefreshLayout = contentView.findViewById(R.id.fragment_message_refresh_layout);
            ThemeUtil.setThemeForSwipeRefreshLayout(swipeRefreshLayout);
            swipeRefreshLayout.setOnRefreshListener(this::refresh);
            recyclerView.setLayoutManager(new MyLinearLayoutManager(context));
            recyclerView.addItemDecoration(new CommonDivider(getAttachContext(), MyLinearLayoutManager.VERTICAL, R.drawable.drawable_divider_1dp, DisplayUtil.dp2px(context, 58), DisplayUtil.dp2px(context, 24)));
            adapter = new MessageListAdapter(context, type);
            adapter.setLoadingView(R.layout.layout_footer_loading);
            adapter.setLoadEndView(R.layout.layout_footer_loadend);
            adapter.setLoadFailedView(R.layout.layout_footer_load_failed);
            adapter.setOnLoadMoreListener((boolean isReload) -> {
                if (isReload) load(false);
                else loadMore();
            });
            recyclerView.setAdapter(adapter);
        }

        boolean needLoad() {
            return dataBean == null;
        }

        public int getType() {
            return type;
        }

        public View getContentView() {
            return contentView;
        }

        public void refresh() {
            swipeRefreshLayout.setRefreshing(true);
            load(true);
        }

        private void load(boolean reload) {
            if (reload) {
                page = 1;
            }
            Callback<MessageListBean> messageListBeanCallback = new Callback<MessageListBean>() {
                @Override
                public void onResponse(@NotNull Call<MessageListBean> call, @NotNull Response<MessageListBean> response) {
                    dataBean = response.body();
                    if (reload) {
                        adapter.reset();
                        adapter.setData(dataBean);
                    } else adapter.addData(dataBean);
                    swipeRefreshLayout.setRefreshing(false);
                }

                @Override
                public void onFailure(@NotNull Call<MessageListBean> call, @NotNull Throwable t) {
                    if (reload) {
                        if (!(t instanceof TiebaException)) {
                            Util.showNetworkErrorSnackbar(recyclerView, () -> refresh());
                            return;
                        }
                    }
                    adapter.loadFailed();
                    swipeRefreshLayout.setRefreshing(false);
                }
            };
            switch (getType()) {
                case TYPE_REPLY_ME:
                    TiebaApi.getInstance().replyMe(page).enqueue(messageListBeanCallback);
                    break;
                case TYPE_AT_ME:
                    TiebaApi.getInstance().atMe(page).enqueue(messageListBeanCallback);
                    break;
            }
        }

        private void loadMore() {
            if (dataBean.getPage().getHasMore().equals("1")) {
                page += 1;
                load(false);
            } else {
                adapter.loadEnd();
            }
        }
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/fragments/MessageFragment.kt
package com.huanchengfly.tieba.post.fragments

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.AppBarLayout.LayoutParams.SCROLL_FLAG_NO_SCROLL
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.activities.FloorActivity
import com.huanchengfly.tieba.post.activities.NewSearchActivity
import com.huanchengfly.tieba.post.activities.ThreadActivity
import com.huanchengfly.tieba.post.adapters.MessageListAdapter
import com.huanchengfly.tieba.post.adapters.TabViewPagerAdapter
import com.huanchengfly.tieba.post.api.TiebaApi
import com.huanchengfly.tieba.post.api.models.MessageListBean
import com.huanchengfly.tieba.post.api.retrofit.doIfFailure
import com.huanchengfly.tieba.post.api.retrofit.doIfSuccess
import com.huanchengfly.tieba.post.api.retrofit.exception.TiebaException
import com.huanchengfly.tieba.post.components.MyLinearLayoutManager
import com.huanchengfly.tieba.post.goToActivity
import com.huanchengfly.tieba.post.interfaces.Refreshable
import com.huanchengfly.tieba.post.utils.ThemeUtil
import com.huanchengfly.tieba.post.utils.Util
import com.scwang.smart.refresh.header.MaterialHeader
import com.scwang.smart.refresh.layout.SmartRefreshLayout

class MessageFragment : BaseFragment(), Refreshable, OnTabSelectedListener,
    Toolbar.OnMenuItemClickListener {
    @BindView(R.id.fragment_message_tab)
    lateinit var tabLayout: TabLayout

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar

    @BindView(R.id.appbar)
    lateinit var mAppBarLayout: AppBarLayout

    @BindView(R.id.fragment_message_vp)
    lateinit var viewPager: ViewPager

    @BindView(R.id.title)
    lateinit var mTitleTextView: TextView

    private val replyMe: MessageListHelper by lazy { MessageListHelper(TYPE_REPLY_ME) }
    private val atMe: MessageListHelper by lazy { MessageListHelper(TYPE_AT_ME) }
    var type = 0
        private set
    private var isFromMain = false
    public override fun onFragmentVisibleChange(isVisible: Boolean) {
        if (isVisible) {
            refreshIfNeed()
        }
    }

    public override fun onFragmentFirstVisible() {
        refreshIfNeed()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val args = arguments
        if (args != null) {
            type = args.getInt(PARAM_TYPE, TYPE_REPLY_ME)
            isFromMain = args.getBoolean(PARAM_FROM_MAIN, false)
        }
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_message
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (!isFromMain) {
            toolbar.visibility = View.GONE
            (toolbar.layoutParams as AppBarLayout.LayoutParams?)?.scrollFlags =
                SCROLL_FLAG_NO_SCROLL
        }
        toolbar.setOnMenuItemClickListener(this)
        viewPager.adapter = TabViewPagerAdapter().apply {
            addView(
                replyMe.requireContentView(),
                attachContext.getString(R.string.title_reply_me)
            )
            addView(
                atMe.requireContentView(),
                attachContext.getString(R.string.title_at_me)
            )
        }
        tabLayout.setupWithViewPager(viewPager)
        tabLayout.addOnTabSelectedListener(this)
        viewPager.setCurrentItem(type, false)
        if (!isFromMain) {
            refreshIfNeed()
        }
    }

    override fun hasOwnAppbar(): Boolean {
        return true
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_search) {
            goToActivity<NewSearchActivity>()
        }
        return false
    }

    override fun onRefresh() {
        when (tabLayout.selectedTabPosition) {
            0 -> replyMe
            1 -> atMe
            else -> null
        }?.apply {
            if (isFragmentVisible) {
                refresh(true)
            } else {
                dataBean = null
            }
        }
    }

    private fun refreshIfNeed() {
        when (tabLayout.selectedTabPosition) {
            0 -> replyMe
            1 -> atMe
            else -> null
        }?.apply {
            if (needLoad()) {
                refresh(true)
            }
        }
    }

    override fun onTabSelected(tab: TabLayout.Tab) {
        refreshIfNeed()
    }

    override fun onTabUnselected(tab: TabLayout.Tab) {}

    override fun onTabReselected(tab: TabLayout.Tab) {
        refreshIfNeed()
    }

    internal inner class MessageListHelper(val type: Int) {
        var contentView: View? = null
        private lateinit var smartRefreshLayout: SmartRefreshLayout
        private lateinit var materialHeader: MaterialHeader
        private lateinit var recyclerView: RecyclerView
        private lateinit var adapter: MessageListAdapter
        private var page = 0
        var dataBean: MessageListBean? = null

        fun needLoad(): Boolean {
            return dataBean == null
        }

        @JvmOverloads
        fun refresh(autoRefresh: Boolean = true) {
            if (autoRefresh) smartRefreshLayout.autoRefresh()
            load(true)
        }

        private fun load(reload: Boolean) {
            if (reload) {
                recyclerView.scrollToPosition(0)
                page = 1
            }
            launchIO {
                when (type) {
                    TYPE_AT_ME -> TiebaApi.getInstance().atMeAsync(page)
                    else -> TiebaApi.getInstance().replyMeAsync(page)
                }.doIfSuccess {
                    dataBean = it
                    if (reload) {
                        adapter.reset()
                        adapter.setData(it)
                        recyclerView.scrollToPosition(0)
                    } else {
                        adapter.addData(it)
                    }
                    smartRefreshLayout.finishRefresh(true)
                    if (it.page?.hasMore != "1") {
                        smartRefreshLayout.finishRefreshWithNoMoreData()
                    }
                }.doIfFailure {
                    if (reload) {
                        if (it !is TiebaException) {
                            Util.showNetworkErrorSnackbar(recyclerView) { refresh() }
                        }
                    }
                    if (reload) {
                        smartRefreshLayout.finishRefresh(false)
                    } else {
                        smartRefreshLayout.finishLoadMore(false)
                    }
                }
            }
        }

        private fun loadMore() {
            if (dataBean!!.page!!.hasMore == "1") {
                page += 1
                load(false)
            } else {
                smartRefreshLayout.finishLoadMoreWithNoMoreData()
            }
        }

        fun requireContentView(): View {
            if (contentView == null) {
                initView()
            }
            return contentView!!
        }

        fun initView() {
            contentView = Util.inflate(attachContext, R.layout.fragment_message_list)!!.also {
                recyclerView = it.findViewById(R.id.fragment_message_recycler_view)
                smartRefreshLayout = it.findViewById(R.id.fragment_message_refresh_layout)
                materialHeader = it.findViewById(R.id.refresh_header)
            }
            ThemeUtil.setThemeForMaterialHeader(materialHeader)
            smartRefreshLayout.setOnRefreshListener {
                refresh(false)
            }
            smartRefreshLayout.setOnLoadMoreListener {
                loadMore()
            }
            recyclerView.layoutManager = MyLinearLayoutManager(context)
            adapter = MessageListAdapter(context!!, type).apply {
                setOnItemClickListener { _, item, _ ->
                    if (item.isFloor == "1") {
                        FloorActivity.launch(
                            attachContext,
                            item.threadId!!,
                            subPostId = item.postId
                        )
                    } else {
                        ThreadActivity.launch(context, item.threadId!!, item.postId)
                    }
                }
            }
            recyclerView.adapter = adapter
        }

        init {
            require(!(this.type != TYPE_REPLY_ME && this.type != TYPE_AT_ME)) { "参数不正确" }
            initView()
        }
    }

    companion object {
        const val TYPE_REPLY_ME = 0
        const val TYPE_AT_ME = 1
        val TAG: String = MessageFragment::class.java.simpleName
        private const val PARAM_TYPE = "type"
        const val PARAM_FROM_MAIN = "from_main"

        @JvmStatic
        @JvmOverloads
        fun newInstance(type: Int, fromMain: Boolean = false): MessageFragment {
            return MessageFragment().apply {
                arguments = Bundle().apply {
                    putInt(PARAM_TYPE, type)
                    putBoolean(PARAM_FROM_MAIN, fromMain)
                }
            }
        }
    }
}